raster/src/main/scala/geotrellis/raster/io/geotiff/OverviewStrategy.scala

```diff
@@ -16,20 +16,95 @@
 
 package geotrellis.raster.io.geotiff
 
+import geotrellis.raster.CellSize
+
+import scala.collection.Searching._
+
+/**
+  * Options ported from GDAL (https://gdal.org/programs/gdalwarp.html#cmdoption-gdalwarp-ovr)
+  *  for automatic selection from among available overviews
+  */
 sealed trait OverviewStrategy
+
+object OverviewStrategy {
+  def DEFAULT = AutoHigherResolution
+
+  /**
+    * Select appropriate overview given the strategy.
+    *
+    * Unless a particular strategy suggests otherwise, this method will clamp the returned
+    * index to the range of overviewCS.
+    * @param overviewCS
+    * @param desiredCS
+    * @param strategy
+    * @return
+    */
+  def selectOverview(
+    overviewCS: List[CellSize],
+    desiredCS: CellSize,
+    strategy: OverviewStrategy
+  ): Int = {
+    val maybeIndex = strategy match {
+      case Level(overviewIdx) =>
+        overviewIdx
+      case Auto(n) =>
+        selectIndexByProximity(overviewCS, desiredCS, 0.5) + n
+      case Base =>
+        overviewCS.indexOf(overviewCS.min)
+      case AutoHigherResolution =>
+        val idx = selectIndexByProximity(overviewCS, desiredCS, 1.0)
+        // AutoHigherResolution defaults to the closest level
+        // idx > overviewCS.size means, that idx = overviewCS.size (last idx + 1)
+        if (idx > overviewCS.size) idx - 1 else idx
+    }
+    if (maybeIndex < 0) 0
+    else if (maybeIndex >= overviewCS.size) overviewCS.size - 1
+    else maybeIndex
+  }
+
+  // This method is unsafe. It's up to the selectOverview method to handle each index OOB error for
+  // each indivdual strategy
+  private def selectIndexByProximity(overviewCS: List[CellSize], desiredCS: CellSize, proximityThreshold: Double): Int =
+    overviewCS.search(desiredCS) match {
```

Comment:
I guess this could be replaed with a [foldM](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/Foldable.scala#L384) function